## Data Generation
* imports and seed

* Importing libraries

In [16]:
import numpy as np
import pandas as pd
from datetime import datetime, timedelta
import os


+ Creating seed

In [15]:
np.random.seed(42)
N_TRANSACTIONS = 5000
os.makedirs('data', exist_ok = True)
os.makedirs('outputs/reports', exist_ok = True)

* Generating Product Catalog

In [21]:
product_names = ['Laptop', 'Wireless Mouse', 'USB-C Hub', 'Monitor', 'Keyboard', 'Webcam', 'Headphones', 'Phone Stand', 'Desk Lamp', 'Notebook']
categories = ['Electronics', 'Electronics', 'Electronics', 'Electronics', 'Electronics', 'Electronics', 'Audio', 'Accessories', 'Accessories', 'Stationery']
base_prices = np.array([999.99, 29.99, 49.99, 349.99, 79.99, 89.99, 149.99, 19.99, 39.99, 9.99])

products_df = pd.DataFrame({'product_id': [f'P{i:03d}' for i in range(1, 11)], 'product_name': product_names, 'category': categories, 'base_price': base_prices})
products_df.to_csv('data/products.csv', index = False)
print('Products saved:', products_df.shape)

Products saved: (10, 4)


* Generating customer profiles

In [10]:
N_CUSTOMERS = 500
cities = ['New York', 'Los Angeles', 'Chicago', 'Houston', 'Phoenix']
tiers = ['Bronze', 'Silver', 'Gold', 'Platinum']
customers_df = pd.DataFrame({
 'customer_id': [f'C{i:04d}' for i in range(1, N_CUSTOMERS + 1)],
 'city': np.random.choice(cities, N_CUSTOMERS),
 'tier': np.random.choice(tiers, N_CUSTOMERS, p=[0.4, 0.3, 0.2, 
0.1]),
 'age': np.random.randint(18, 75, N_CUSTOMERS),
 'join_year': np.random.randint(2018, 2024, N_CUSTOMERS)
})
customers_df.to_csv('data/customers.csv', index=False)
print('Customers saved:', customers_df.shape)

Customers saved: (500, 5)


* Generating Sales Transactions

In [11]:
start_date = datetime(2023, 1, 1)
date_offsets = np.random.randint(0, 365, N_TRANSACTIONS)
dates = [start_date + timedelta(days=int(d)) for d in date_offsets]
product_ids = np.random.choice(products_df['product_id'], 
N_TRANSACTIONS)
customer_ids = np.random.choice(customers_df['customer_id'], 
N_TRANSACTIONS)

quantities = np.random.choice(
 [1, 2, 3, 4, 5], N_TRANSACTIONS, p=[0.5, 0.25, 0.12, 0.08, 0.05]
)
discount_pct = np.random.choice(
 [0, 5, 10, 15, 20], N_TRANSACTIONS, p=[0.5, 0.2, 0.15, 0.1, 0.05]
)
sales_df = pd.DataFrame({
 'transaction_id': [f'T{i:05d}' for i in range(1, N_TRANSACTIONS + 
1)],
 'date': dates,
 'product_id': product_ids,
 'customer_id': customer_ids,
 'quantity': quantities,
 'discount_pct': discount_pct
})
# Injecting ~2% nulls for cleaning practice
null_idx = np.random.choice(sales_df.index, size=100, replace=False)
sales_df.loc[null_idx[:50], 'quantity'] = np.nan
sales_df.loc[null_idx[50:], 'discount_pct'] = np.nan
sales_df.to_csv('data/sales_data.csv', index=False)
print('Sales saved:', sales_df.shape)

Sales saved: (5000, 6)


* Initial inspection

In [12]:
print('\n--- SALES DATA PREVIEW ---')
print(sales_df.head(10).to_string())
print('\n--- COLUMN TYPES ---')
print(sales_df.dtypes)
print('\n--- MISSING VALUES ---')
print(sales_df.isnull().sum())
print('\n--- DESCRIPTIVE STATS ---')
print(sales_df.describe())


--- SALES DATA PREVIEW ---
  transaction_id       date product_id customer_id  quantity  discount_pct
0         T00001 2023-12-30       P004       C0026       1.0           5.0
1         T00002 2023-01-21       P010       C0034       1.0          20.0
2         T00003 2023-04-17       P006       C0264       1.0           0.0
3         T00004 2023-12-02       P002       C0138       1.0           5.0
4         T00005 2023-10-30       P005       C0304       4.0           0.0
5         T00006 2023-04-27       P006       C0183       1.0           5.0
6         T00007 2023-05-04       P005       C0363       1.0           0.0
7         T00008 2023-10-08       P002       C0242       5.0           0.0
8         T00009 2023-10-19       P007       C0054       1.0           5.0
9         T00010 2023-07-22       P004       C0462       3.0           0.0

--- COLUMN TYPES ---
transaction_id               str
date              datetime64[us]
product_id                   str
customer_id               

np.random.seed(42) inputs the number 42 into the mersenne twister algorithm and generates a unique sequence of 'random' numbers from it.  
This allows one to get the same sequence every time they run the code, making it pseudorandom.  
Removing the nummber makes the sequence change every time you run it, preventing consistent results.

In [13]:
sales_df.describe()

,date,quantity,discount_pct
count,5000,4950.000000,4950.000000
mean,2023-06-30 03:58:10.560000,1.936970,5.211111
min,2023-01-01 00:00:00,1.000000,0.000000
25%,2023-04-03 00:00:00,1.000000,0.000000
50%,2023-06-28 00:00:00,1.000000,5.000000
75%,2023-09-29 06:00:00,3.000000,10.000000
max,2023-12-31 00:00:00,5.000000,20.000000
std,NaN,1.194242,6.183109


The mean quantity is 1.936970  
The mean discount is 5.211111

10000 transactions could be generated by changing the value of N_TRANSACTIONS from 5000 to 10000.  


In [22]:
products_df.sort_values(by = 'base_price', ascending = False)

,product_id,product_name,category,base_price
0,P001,Laptop,Electronics,999.99
3,P004,Monitor,Electronics,349.99
6,P007,Headphones,Audio,149.99
5,P006,Webcam,Electronics,89.99
4,P005,Keyboard,Electronics,79.99
2,P003,USB-C Hub,Electronics,49.99
8,P009,Desk Lamp,Accessories,39.99
1,P002,Wireless Mouse,Electronics,29.99
7,P008,Phone Stand,Accessories,19.99
9,P010,Notebook,Stationery,9.99


The product with the highest base price is the Laptop

# Data Cleaning and Preprocessing
* Loading and Inspecting

In [3]:
import numpy as np
import pandas as pd

sales = pd.read_csv('data/sales_data.csv')
products = pd.read_csv('data/products.csv')
customers = pd.read_csv('data/customers.csv')
print('Shape:', sales.shape)
print(sales.dtypes)
print('Nulls:\n', sales.isnull().sum())

Shape: (5000, 6)
transaction_id        str
date                  str
product_id            str
customer_id           str
quantity          float64
discount_pct      float64
dtype: object
Nulls:
 transaction_id     0
date               0
product_id         0
customer_id        0
quantity          50
discount_pct      50
dtype: int64


+ Fixing date type and engineer features

In [4]:
sales['date'] = pd.to_datetime(sales['date'])
sales['month'] = sales['date'].dt.month
sales['month_name'] = sales['date'].dt.strftime('%b') # .strftime() stands for 'string format time'
sales['quarter'] = sales['date'].dt.quarter
sales['day_of_week'] = sales['date'].dt.day_name()
sales['week_number'] = sales['date'].dt.isocalendar().week.astype(int)

+ Handling missing values

In [5]:
# Filling quantity with median 
median_qty = sales['quantity'].median()
sales['quantity'] = sales['quantity'].fillna(median_qty).astype(int)
# Filling discount_pct with 0 (assumes no discount if unknown)
sales['discount_pct'] = sales['discount_pct'].fillna(0)
print('Nulls after fill:', sales.isnull().sum().sum())

Nulls after fill: 0


* Merging all three datasets


In [9]:
# Left join: keep all sales rows
df = sales.merge(products, on='product_id', how='left')
df = df.merge(customers, on='customer_id', how='left')
print('Master shape:', df.shape)
print('Columns:', df.columns.tolist())

Master shape: (5000, 18)
Columns: ['transaction_id', 'date', 'product_id', 'customer_id', 'quantity', 'discount_pct', 'month', 'month_name', 'quarter', 'day_of_week', 'week_number', 'product_name', 'category', 'base_price', 'city', 'tier', 'age', 'join_year']


* Computing revenue

In [ ]:
# Revenue = quantity x base_price x (1 - discount/100)
df['revenue'] = (
 df['quantity'] * df['base_price'] *
 (1 - df['discount_pct'] / 100)
).round(2)

* Detecting outliers with NumPy IQR

In [14]:
revenue_arr = df['revenue'].values
Q1 = np.percentile(revenue_arr, 25)
Q3 = np.percentile(revenue_arr, 75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outlier_mask = (df['revenue'] < lower_bound) | (df['revenue'] > 
upper_bound)
print(f'Outliers: {outlier_mask.sum()} rows ({outlier_mask.mean()*100:.1f}%)')
# Cap outliers instead of dropping
df['revenue_capped'] = df['revenue'].clip(lower_bound, upper_bound)

Outliers: 710 rows (14.2%)


* Validating and saving

In [15]:
assert df.isnull().sum().sum() == 0, 'Nulls remain!'
assert (df['revenue'] >= 0).all(), 'Negative revenue!'
assert not df.duplicated('transaction_id').any(), 'Duplicate IDs!'
print('All validation checks passed.')
df.to_csv('data/master_data.csv', index=False)

All validation checks passed.


fillna() fills any null values in a row/column with the selected value   while dropna() removes any row/column that has a null value. dropna is used when the user is okay with deleting entire rows from their data

Median isn't easily affected by ouliers unlike the mean value, so it's used when filling NaN values

left merge includes everything in the left data frame while also including what is common between both data frames.  The number of rows in the result was 5000

In [25]:
b = df['discount_pct'] > 0
b.sum()


np.int64(2574)

there are 2574 transactions with discounts greater than 0

# NumPy Deep Dive — Statistical Analysis
* Loading data as NumPy arrays

In [2]:
df = pd.read_csv('data/master_data.csv', parse_dates=['date'])
# Extracting NumPy arrays
revenue = df['revenue'].values
quantity = df['quantity'].values
prices = df['base_price'].values
discounts = df['discount_pct'].values
months = df['month'].values

* Full Descriptive Statistics

In [28]:
def describe_array(arr, label):
 print(f'\n=== {label} ===')
 print(f' Count : {len(arr):>10,}')
 print(f' Mean : {np.mean(arr):>10.2f}')
 print(f' Median : {np.median(arr):>10.2f}')
 print(f' Std Dev : {np.std(arr):>10.2f}')
 print(f' Min : {np.min(arr):>10.2f}')
 print(f' Max : {np.max(arr):>10.2f}')
 print(f' Range : {np.ptp(arr):>10.2f}')
 print(f' P25 : {np.percentile(arr, 25):>10.2f}')
 print(f' P75 : {np.percentile(arr, 75):>10.2f}')
 print(f' P90 : {np.percentile(arr, 90):>10.2f}')
describe_array(revenue, 'REVENUE ($)')
describe_array(quantity, 'QUANTITY (units)')
describe_array(discounts, 'DISCOUNT (%)')


=== REVENUE ($) ===
 Count :      5,000
 Mean :     332.90
 Median :      89.99
 Std Dev :     654.91
 Min :       7.99
 Max :    4999.95
 Range :    4991.96
 P25 :      39.98
 P75 :     299.98
 P90 :     949.99

=== QUANTITY (units) ===
 Count :      5,000
 Mean :       1.93
 Median :       1.00
 Std Dev :       1.19
 Min :       1.00
 Max :       5.00
 Range :       4.00
 P25 :       1.00
 P75 :       3.00
 P90 :       4.00

=== DISCOUNT (%) ===
 Count :      5,000
 Mean :       5.16
 Median :       5.00
 Std Dev :       6.17
 Min :       0.00
 Max :      20.00
 Range :      20.00
 P25 :       0.00
 P75 :      10.00
 P90 :      15.00


* Monthly Revenue Aggregation

In [29]:
monthly_revenue = np.array([
 revenue[months == m].sum() for m in range(1, 13)
])
month_names = ['Jan','Feb','Mar','Apr','May','Jun',
 'Jul','Aug','Sep','Oct','Nov','Dec']
print('\n=== MONTHLY REVENUE ===')
for i, (name, rev) in enumerate(zip(month_names, monthly_revenue), 1):
 bar = '#' * int(rev / monthly_revenue.max() * 30)
 print(f' {name}: ${rev:>10,.2f} {bar}')
best_month_idx = np.argmax(monthly_revenue)
worst_month_idx = np.argmin(monthly_revenue)

print(f'\nBest month: {month_names[best_month_idx]} (${monthly_revenue[best_month_idx]:,.2f})')
print(f'Worst month: {month_names[worst_month_idx]} (${monthly_revenue[worst_month_idx]:,.2f})')


=== MONTHLY REVENUE ===
 Jan: $144,256.82 ###########################
 Feb: $133,177.33 #########################
 Mar: $132,731.94 #########################
 Apr: $150,001.68 ############################
 May: $140,139.65 ##########################
 Jun: $135,980.02 ##########################
 Jul: $142,311.30 ###########################
 Aug: $127,776.13 ########################
 Sep: $155,951.66 ##############################
 Oct: $122,780.37 #######################
 Nov: $135,700.87 ##########################
 Dec: $143,692.67 ###########################

Best month: Sep ($155,951.66)
Worst month: Oct ($122,780.37)


* 3-Month Rolling Average

In [30]:
window = 3
weights = np.ones(window) / window
rolling_avg = np.convolve(monthly_revenue, weights, mode='valid')
print('\n=== 3-MONTH ROLLING AVERAGE ===')
for i, avg in enumerate(rolling_avg):
 print(f' {month_names[i+1]}: ${avg:>10,.2f}')


=== 3-MONTH ROLLING AVERAGE ===
 Feb: $136,722.03
 Mar: $138,636.98
 Apr: $140,957.76
 May: $142,040.45
 Jun: $139,476.99
 Jul: $135,355.82
 Aug: $142,013.03
 Sep: $135,502.72
 Oct: $138,144.30
 Nov: $134,057.97


* Broadcasting: Tier Multiplier

In [3]:
tier_map = {'Bronze': 1.0, 'Silver': 1.05, 'Gold': 1.10, 'Platinum': 1.15}
multiplier = df['tier'].map(tier_map).values
# NumPy broadcasting — no Python loops
adjusted_revenue = revenue * multiplier
print(f'\nOriginal total revenue : ${revenue.sum():>12,.2f}')
print(f'Tier-adjusted revenue : ${adjusted_revenue.sum():>12,.2f}')
print(f'Tier uplift : ${adjusted_revenue.sum()-revenue.sum():>12,.2f}')


Original total revenue : $1,664,500.44
Tier-adjusted revenue : $1,740,382.31
Tier uplift : $   75,881.87


* Correlation Matrix

In [9]:
corr = np.corrcoef([prices, quantity, discounts])
labels = ['Price', 'Quantity', 'Discount']
print('\n=== PEARSON CORRELATION MATRIX ===')
header = f' {'':12s}' + ''.join(f'{l:>12s}' for l in labels)
print(header)
for i, row in enumerate(corr):
 print(f' {labels[i]:12s}' + ''.join(f'{v:>12.3f}' for v in row))


=== PEARSON CORRELATION MATRIX ===
                    Price    Quantity    Discount
 Price              1.000      -0.004      -0.019
 Quantity          -0.004       1.000       0.008
 Discount          -0.019       0.008       1.000


* Top 10 Highest Revenue Transactions

In [10]:
top10_idx = np.argsort(revenue)[-10:][::-1]
print('\n=== TOP 10 REVENUE TRANSACTIONS ===')
for rank, idx in enumerate(top10_idx, 1):
 tid = df.iloc[idx]['transaction_id']
 rev = revenue[idx]
 prod = df.iloc[idx]['product_name']
 print(f' {rank:>2}. {tid} {prod:20s} ${rev:>8,.2f}')


=== TOP 10 REVENUE TRANSACTIONS ===
  1. T04661 Laptop               $4,999.95
  2. T00468 Laptop               $4,999.95
  3. T02813 Laptop               $4,999.95
  4. T03204 Laptop               $4,999.95
  5. T02992 Laptop               $4,999.95
  6. T02844 Laptop               $4,999.95
  7. T02413 Laptop               $4,999.95
  8. T02755 Laptop               $4,999.95
  9. T04508 Laptop               $4,999.95
 10. T01583 Laptop               $4,999.95


* Flag high-value transactions with np.where

In [14]:
threshold = np.percentile(revenue, 90)
df['value_segment'] = np.where(revenue >= threshold, 'High-Value', 
'Standard')
high_val_count = (df['value_segment'] == 'High-Value').sum()
high_val_rev = revenue[revenue >= threshold].sum()
print(f'\nHigh-value transactions (top 10%): {high_val_count}')
print(f'Revenue from high-value transactions: ${high_val_rev:,.2f}')


High-value transactions (top 10%): 547
Revenue from high-value transactions: $1,004,042.10


np.mean() calculates average of the column i.e sum of the values divided by the number of values    
np.median() returns the middle value gotten when the column is sorted from the lowest to highest value    
Median is more appropriate because it isn't easily affected by outliers compared to the mean

np.convolve() rolls the window over the array and calculates the product of the sum of the values
 
 by using mode = valid, the window is applied to only the parts of the array it can fit in, its used when you want to avoid getting data
for periods that don't exist

there is no meaningful correlation between them  
 the value being 0.008 suggests that they had little to no impact on each other   
 
this is due to the fact a random number generator was used in making the data 

In [17]:
total = revenue.sum()
top10rev = (high_val_rev / total) * 100
top10rev


np.float64(60.320927280740165)

60% of the total revenue comes from the top ten transactions

# Pandas Deep Dive — Business Insights
* Revenue and Volume by Category

In [18]:
import pandas as pd
import numpy as np
import os
df = pd.read_csv('data/master_data.csv', parse_dates=['date'])
os.makedirs('outputs', exist_ok=True)

category_summary = (
 df.groupby('category')
 .agg(
 total_revenue = ('revenue', 'sum'),
 avg_order_value = ('revenue', 'mean'),
 total_units = ('quantity', 'sum'),
 num_transactions = ('transaction_id', 'count'),
 unique_customers = ('customer_id', 'nunique'),
 avg_discount = ('discount_pct', 'mean')
 )
 .sort_values('total_revenue', ascending=False)
 .round(2)
)
# Add revenue share column
category_summary['revenue_share_pct'] = (
 category_summary['total_revenue'] /
 category_summary['total_revenue'].sum() * 100
).round(1)
print(category_summary.to_string())
category_summary.to_csv('outputs/category_summary.csv')

             total_revenue  avg_order_value  total_units  num_transactions  unique_customers  avg_discount  revenue_share_pct
category                                                                                                                     
Electronics     1459103.89           493.77         5667              2955               498          5.08               87.7
Audio            139557.87           272.57          979               512               334          5.18                8.4
Accessories       55734.18            56.64         1921               984               436          5.14                3.3
Stationery        10104.50            18.41         1071               549               340          5.61                0.6


* Top 10 Customers by Lifetime Value

In [19]:
customer_ltv = (
 df.groupby(['customer_id', 'city', 'tier'])
 .agg(
 lifetime_value = ('revenue', 'sum'),
 total_orders = ('transaction_id', 'count'),
 avg_order_value = ('revenue', 'mean'),
 total_units = ('quantity', 'sum'),
 first_purchase = ('date', 'min'),
 last_purchase = ('date', 'max'),
 )
 .reset_index()
 .sort_values('lifetime_value', ascending=False)
 .head(10)
 .round(2)
)
print(customer_ltv[['customer_id','city','tier',
 
'lifetime_value','total_orders']].to_string(index=False))
customer_ltv.to_csv('outputs/top10_customers.csv', index=False)

customer_id        city   tier  lifetime_value  total_orders
      C0444     Chicago Silver        17791.12            15
      C0357     Houston   Gold        14480.63            17
      C0432    New York Bronze        13049.67            17
      C0494    New York Bronze        11317.24            11
      C0216     Chicago Bronze        10981.80             9
      C0042    New York Bronze        10884.25             9
      C0016 Los Angeles Bronze        10272.60            18
      C0348    New York Silver         9876.83            10
      C0456 Los Angeles Silver         9775.76            11
      C0275     Chicago Bronze         9676.78            12


C:\Users\AB\AppData\Local\Temp\ipykernel_12264\971522423.py:14: UserWarning: obj.round has no effect with datetime, timedelta, or period dtypes. Use obj.dt.round(...) instead.
  .round(2)


* Month-over-Month Revenue Trend 

In [21]:
monthly = (df.set_index('date').resample('ME').agg(monthly_revenue = ('revenue', 'sum'),num_transactions = ('transaction_id', 'count'),
 avg_order_value = ('revenue', 'mean'),
 total_units = ('quantity', 'sum')))
monthly['mom_change_pct'] = monthly['monthly_revenue'].pct_change() * 100
monthly['cum_revenue'] = monthly['monthly_revenue'].cumsum()
monthly['rolling_3m_avg'] = monthly['monthly_revenue'].rolling(3).mean()
monthly = monthly.round(2)
print(monthly.to_string())
monthly.to_csv('outputs/monthly_trend.csv')

            monthly_revenue  num_transactions  avg_order_value  total_units  mom_change_pct  cum_revenue  rolling_3m_avg
date                                                                                                                    
2023-01-31        144256.82               427           337.84          832             NaN    144256.82             NaN
2023-02-28        133177.33               373           357.04          732           -7.68    277434.15             NaN
2023-03-31        132731.94               430           308.68          818           -0.33    410166.09       136722.03
2023-04-30        150001.68               447           335.57          852           13.01    560167.77       138636.98
2023-05-31        140139.65               448           312.81          854           -6.57    700307.42       140957.76
2023-06-30        135980.02               414           328.45          819           -2.97    836287.44       142040.45
2023-07-31        142311.30     

* Tier vs Average Order Value (Pivot Table)

In [ ]:
tier_pivot = pd.pivot_table(
    df,
    values='revenue',
    index='tier',
    columns='quarter',
    aggfunc=['mean', 'sum', 'count'], 
    margins=True,
    margins_name='All Tiers'
).round(2)

print(tier_pivot.to_string())
tier_pivot.to_csv('outputs/tier_pivot.csv')

             mean                                          sum                                              count                            
quarter         1       2       3       4 All Tiers          1          2          3          4   All Tiers     1     2     3     4 All Tiers
tier                                                                                                                                         
Bronze     334.34  327.32  353.43  356.51    342.70  176533.10  183301.98  195095.03  183245.95   738176.06   528   560   552   514      2154
Gold       293.09  299.82  349.46  315.69    313.13   69461.37   72256.80   71638.75   63138.23   276495.15   237   241   205   200       883
Platinum   353.14  223.76  286.88  335.66    300.40   49440.17   30431.49   33565.25   43972.00   157408.91   140   136   117   131       524
Silver     353.02  376.70  347.35  294.26    342.20  114731.45  140131.08  125740.06  111817.73   492420.32   325   372   362   380      1439
All Ti

* Best Product per City

In [25]:
city_product_rev = (
 df.groupby(['city', 'product_name'])['revenue']
 .sum()
 .unstack(fill_value=0)
)
best_product_per_city = pd.DataFrame({
 'best_product' : city_product_rev.idxmax(axis=1),
 'top_revenue' : city_product_rev.max(axis=1).round(2),
 'num_products_sold': (city_product_rev > 0).sum(axis=1)
})
print(best_product_per_city.to_string())
best_product_per_city.to_csv('outputs/best_product_per_city.csv')

            best_product  top_revenue  num_products_sold
city                                                    
Chicago           Laptop    175598.19                 10
Houston           Laptop    221097.72                 10
Los Angeles       Laptop    155898.40                 10
New York          Laptop    202447.91                 10
Phoenix           Laptop    170398.23                 10


* Discount Level Impact on Revenue

In [40]:
# Bin discount levels
df['discount_band'] = pd.cut(
 df['discount_pct'],
 bins=[-1, 0, 5, 10, 15, 20],
 labels=['No Discount', '1-5%', '6-10%', '11-15%', '16-20%']
)
discount_impact = (
 df.groupby('discount_band', observed=True)
 .agg(
 num_transactions = ('transaction_id', 'count'),
 total_revenue = ('revenue', 'sum'),
 avg_revenue = ('revenue', 'mean'),
 avg_quantity = ('quantity', 'mean'),
 )
 .round(2)
)
print(discount_impact.to_string())
discount_impact.to_csv('outputs/discount_impact.csv')

               num_transactions  total_revenue  avg_revenue  avg_quantity
discount_band                                                            
No Discount                2426      856273.38       352.96          1.92
1-5%                       1028      340174.88       330.91          1.96
6-10%                       776      252201.88       325.00          1.87
11-15%                      501      143063.00       285.55          1.92
16-20%                      269       72787.30       270.58          2.04


* Product Performance Matrix

In [29]:
product_matrix = (
 df.groupby(['product_name', 'category'])
 .agg(
 units_sold = ('quantity', 'sum'),
 total_revenue = ('revenue', 'sum'),
 avg_discount = ('discount_pct', 'mean'),
 unique_customers = ('customer_id', 'nunique'),
 num_transactions = ('transaction_id', 'count'),
 )
 .sort_values('total_revenue', ascending=False)
 .round(2)
)
product_matrix['revenue_per_unit'] = (
 product_matrix['total_revenue'] / product_matrix['units_sold']
).round(2)
print(product_matrix.to_string())
product_matrix.to_csv('outputs/product_matrix.csv')

                            units_sold  total_revenue  avg_discount  unique_customers  num_transactions  revenue_per_unit
product_name   category                                                                                                  
Laptop         Electronics         975      925440.45          4.87               328               511            949.17
Monitor        Electronics         935      310685.81          4.97               321               479            332.28
Headphones     Audio               979      139557.87          5.18               334               512            142.55
Webcam         Electronics         938       80117.81          5.04               308               499             85.41
Keyboard       Electronics         941       70970.78          5.50               302               486             75.42
USB-C Hub      Electronics         970       46075.48          4.96               317               490             47.50
Desk Lamp      Accessori

pct_change() is what calculated the month over month revenue, used in checking the growth over a month. 

it sums the current and previous revenue, divides the sum by the previous and multiplies it by 100 percent. 

The months with negative growth were february, march, may, june, august and october

In [39]:
b = product_matrix
b['AOV'] = b['total_revenue'] / b['num_transactions']
b

,,units_sold,total_revenue,avg_discount,unique_customers,num_transactions,revenue_per_unit,AOV
product_name,category,,,,,,,
Laptop,Electronics,975,925440.45,4.87,328,511,949.17,1811.038063
Monitor,Electronics,935,310685.81,4.97,321,479,332.28,648.613382
Headphones,Audio,979,139557.87,5.18,334,512,142.55,272.573965
Webcam,Electronics,938,80117.81,5.04,308,499,85.41,160.556733
Keyboard,Electronics,941,70970.78,5.50,302,486,75.42,146.030412
USB-C Hub,Electronics,970,46075.48,4.96,317,490,47.50,94.031592
Desk Lamp,Accessories,1015,38582.07,4.96,333,514,38.01,75.062393
Wireless Mouse,Electronics,908,25813.56,5.14,314,490,28.43,52.680735
Phone Stand,Accessories,906,17152.11,5.33,306,470,18.93,36.493851


The Electronics category has the highes average order value

The platinum teir does not consistently out perform the bronze teir across all quarters

The average quantity purchased peaked at discounts between 16% and 20%

# Report Generation & Excel Export
+ Setup and helpers

In [41]:
import pandas as pd
import numpy as np
from tabulate import tabulate
from openpyxl import load_workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter
import os
import datetime
# Load master data and all pre-computed outputs
df = pd.read_csv('data/master_data.csv', 
parse_dates=['date'])
category_sum = pd.read_csv('outputs/category_summary.csv', 
index_col=0)
top_customers = pd.read_csv('outputs/top10_customers.csv')
monthly = pd.read_csv('outputs/monthly_trend.csv', index_col=0)
product_matrix = pd.read_csv('outputs/product_matrix.csv')
discount_impact = pd.read_csv('outputs/discount_impact.csv')
best_per_city = pd.read_csv('outputs/best_product_per_city.csv', 
index_col=0)
os.makedirs('outputs/reports', exist_ok=True)

* Helper: ASCII section divider

In [44]:
def section(title, width=60):
 line = '=' * width
 return f'\n{line}\n {title}\n{line}'

* Executive Summary

In [45]:
total_revenue = df['revenue'].sum()
total_transactions = len(df)
avg_order_value = df['revenue'].mean()
total_units = df['quantity'].sum()
unique_customers = df['customer_id'].nunique()
best_category = category_sum['total_revenue'].idxmax()
best_month_idx = monthly['monthly_revenue'].idxmax()
best_customer_id = top_customers.iloc[0]['customer_id']
best_customer_ltv = top_customers.iloc[0]['lifetime_value']
# Month-over-month changes
mom = monthly['mom_change_pct'].dropna()
positive_months = (mom > 0).sum()
negative_months = (mom < 0).sum()
best_growth_month = mom.idxmax()
worst_growth_month = mom.idxmin()
report_date = datetime.date.today().strftime('%B %d, %Y')
exec_summary = f'''SHOPSMART INC. — ANNUAL SALES REPORT 2023
Report Generated: {report_date}
Prepared by: Data Analytics Team
{section('FINANCIAL HIGHLIGHTS')}
 Total Revenue : ${total_revenue:>12,.2f}
 Total Transactions : {total_transactions:>12,}
 Total Units Sold : {total_units:>12,}
 Average Order Value : ${avg_order_value:>12,.2f}
 Unique Customers : {unique_customers:>12,}
{section('TOP PERFORMERS')}
 Best Category : {best_category}
 Best Month : {best_month_idx}
 Top Customer : {best_customer_id} (LTV: 
${best_customer_ltv:,.2f})
{section('MONTH-OVER-MONTH GROWTH')}
 Months with growth : {positive_months} of 11
 Months with decline : {negative_months} of 11
 Strongest growth month: {best_growth_month}
 Weakest growth month : {worst_growth_month}
{section('REVENUE BY CATEGORY')}
'''
# Append category table
cat_display = category_sum[['total_revenue','revenue_share_pct','num_transactions']].copy()
cat_display.columns = ['Total Revenue ($)', 'Share (%)', 
'Transactions']
exec_summary += tabulate(cat_display, headers='keys', tablefmt='grid',
 floatfmt='.2f') + '\n'
# Save report
with open('outputs/reports/executive_summary.txt', 'w') as f:
 f.write(exec_summary)
print('Executive summary saved.')

Executive summary saved.


* Category Detailed Report

In [46]:
cat_report = section('SHOPSMART INC. — CATEGORY PERFORMANCE REPORT 2023')
cat_report += '\n\n1. FULL CATEGORY BREAKDOWN\n'
cat_report += tabulate(
 category_sum.reset_index(),
 headers=['Category','Revenue','Avg Order','Units','Transactions',
 'Customers','Avg Discount','Share %'],
 tablefmt='grid', floatfmt='.2f'
) + '\n'
# Product-level breakdown
cat_report += '\n2. PRODUCT PERFORMANCE WITHIN CATEGORIES\n'
for cat, grp in product_matrix.groupby('category'):
 cat_report += f'\n [{cat}]\n'
 sub = grp[['product_name','units_sold','total_revenue',
 'revenue_per_unit','unique_customers']].copy()
 cat_report += tabulate(sub, headers='keys', tablefmt='simple',
 showindex=False, floatfmt='.2f') + '\n'
with open('outputs/reports/category_report.txt', 'w') as f:
 f.write(cat_report)
print('Category report saved.')

Category report saved.


+ Customer Report

In [47]:
cust_report = section('SHOPSMART INC. — CUSTOMER REPORT 2023')
cust_report += '\n\nTOP 10 CUSTOMERS BY LIFETIME VALUE\n'
cust_report += tabulate(
 top_customers[['customer_id','city','tier','lifetime_value',
 'total_orders','avg_order_value']],
 headers=['ID','City','Tier','LTV ($)','Orders','Avg Order ($)'],
 tablefmt='grid', showindex=False, floatfmt='.2f'
) + '\n'
# Tier summary
tier_summary = df.groupby('tier').agg(
 customers = ('customer_id', 'nunique'),
 total_rev = ('revenue', 'sum'),
 avg_order = ('revenue', 'mean'),
 total_orders= ('transaction_id', 'count')
).round(2)
cust_report += '\nCUSTOMER TIER SUMMARY\n'
cust_report += tabulate(tier_summary, headers='keys',
 tablefmt='grid', floatfmt='.2f') + '\n'
with open('outputs/reports/customer_report.txt', 'w') as f:
 f.write(cust_report)
print('Customer report saved.')

Customer report saved.


+ Multi-Sheet Workbook

In [49]:
excel_path = 'outputs/shopsmart_annual_report.xlsx'
with pd.ExcelWriter(excel_path, engine='openpyxl') as writer:
 # Sheet 1: Executive KPIs
 kpi_data = pd.DataFrame({'Metric': ['Total Revenue', 'Total Transactions', 'Total Units',
 'Avg Order Value', 'Unique Customers', 'Best Category',
 'Best Month', 'Top Customer'],
 'Value': [f'${total_revenue:,.2f}', f'{total_transactions:,}',
 f'{total_units:,}', f'${avg_order_value:.2f}',
 f'{unique_customers:,}', best_category,
 str(best_month_idx), best_customer_id]
 })
 kpi_data.to_excel(writer, sheet_name='Executive Summary', 
index=False)
 # Sheet 2: Category Summary
 category_sum.reset_index().to_excel(
 writer, sheet_name='Category Analysis', index=False
 )
 # Sheet 3: Top Customers
 top_customers.to_excel(writer, sheet_name='Top Customers', 
index=False)
 # Sheet 4: Monthly Trend
 monthly.reset_index().to_excel(writer, sheet_name='Monthly Trend', 
index=False)
 # Sheet 5: Product Matrix
 product_matrix.reset_index().to_excel(
 writer, sheet_name='Product Performance', index=False
 )
 # Sheet 6: Discount Analysis
 discount_impact.reset_index().to_excel(
 writer, sheet_name='Discount Analysis', index=False
 )

+ Applying Excel formatting with openpyxl

In [53]:
wb = load_workbook(excel_path)
header_fill = PatternFill('solid', fgColor='2B5797')
header_font = Font(bold=True, color='FFFFFF', size=11)
alt_fill = PatternFill('solid', fgColor='EEF3FB')
thin = Side(border_style='thin', color='CCCCCC')
cell_border = Border(left=thin, right=thin, top=thin, bottom=thin)
for sheet_name in wb.sheetnames:
 ws = wb[sheet_name]
 # Style header row
 for cell in ws[1]:
  cell.fill = header_fill
  cell.font = header_font
  cell.alignment = Alignment(horizontal='center', 
  vertical='center')
  cell.border = cell_border
 # Alternate row shading and borders for data rows
 for row_idx, row in enumerate(ws.iter_rows(min_row=2), start=2):
  fill = alt_fill if row_idx % 2 == 0 else PatternFill()
  for cell in row:
    cell.fill = fill
    cell.border = cell_border
 # Auto-size columns
 for col in ws.columns:
  max_len = max(
  len(str(c.value)) if c.value else 0 for c in col)
  ws.column_dimensions[get_column_letter(col[0].column)].width = min(max_len + 4, 40)
wb.save(excel_path)
print(f'Excel workbook saved: {excel_path}')

Excel workbook saved: outputs/shopsmart_annual_report.xlsx


+ Final console summary

In [54]:
print(section('PROJECT COMPLETE — FINAL SUMMARY', width=70))
print(f' Text Reports : outputs/reports/ (3 files)')
print(f' Excel Report : outputs/shopsmart_annual_report.xlsx')
print(f' Sheets : {len(wb.sheetnames)} sheets with formatted data')
print(f' Total Revenue : ${total_revenue:,.2f}')
print(f' Transactions : {total_transactions:,}')
print(f' Avg Order : ${avg_order_value:.2f}')
print('=' * 70)


 PROJECT COMPLETE — FINAL SUMMARY
 Text Reports : outputs/reports/ (3 files)
 Excel Report : outputs/shopsmart_annual_report.xlsx
 Sheets : 6 sheets with formatted data
 Total Revenue : $1,664,500.44
 Transactions : 5,000
 Avg Order : $332.90
